SPEC.md

```diff
@@ -0,0 +1,334 @@
+
+- [Service Name](#service-name)
+- [Operation Name](#operation-name)
+- [Operation ID](#operation-id)
+
+> **Note:** More details regarding IDs will be provided upon finalizing the ID proposal.
+
+### Service Name
+
+Service names:
+
+- MUST not be empty.
+- MUST consist of only unreserved URL characters as delineated in [RFC3986 Section 2.3][rfc3986-section-2.3].
+
+Valid characters include:
+
+```
+ALPHA / DIGIT / "-" / "." / "_" / "~"
+```
+
+### Operation Name
+
+Operation names:
+
+- MUST not be empty.
+- MUST consist of only unreserved URL characters as delineated in [RFC3986 Section 2.3][rfc3986-section-2.3].
+
+Valid characters include:
+
+```
+ALPHA / DIGIT / "-" / "." / "_" / "~"
+```
+
+### Operation ID
+
+Operation IDs:
+
+- MUST not be empty.
+- MUST consist of only unreserved URL characters as delineated in [RFC3986 Section 2.3][rfc3986-section-2.3].
+
+Valid characters include:
+
+```
+ALPHA / DIGIT / "-" / "." / "_" / "~"
+```
+
+## Schema Definitions
+
+### Failure
+
+The `Failure` object represents outcomes that are `failed` or `canceled`. The object MUST adhere to the following JSON
+schema:
+
+```yaml
+type: object
+properties:
+  message:
+    type: string
+  details:
+    type: object
+    properties:
+      metadata:
+        type: object
+        additionalProperties:
+          type: string
+        description: |
+          String to string mapping, may contain information for decoding the data field.
+      data:
+        type: string
+        description: |
+          Arbitrary string-encoded data; use encodings like base64 for binary data transmission.
+```
+
+### OperationStarted
+
+An `OperationStarted` object indicates the commencement of an asynchronous operation. This object MUST follow the given
+JSON schema:
+
+```yaml
+type: object
+properties:
+  operationId:
+    type: string
+    description: |
+      An identifier for referencing this operation. If provided in the start request, it is echoed back.
+  callbackUrlSupported:
+    type: boolean
+    description: |
+      If the request specified a callback_url and the handler supports callbacks for this operation, this flag will be
+      set.
+      It is up to the caller to decided how to get the outcome of this call in case the handler does not support
+      callbacks.
+```
+
+### OperationInfo
+
+The `OperationInfo` object, retrieved from the get operation info API, MUST adhere to the given schema:
+
+```yaml
+type: object
+properties:
+  state:
+    enum:
+      - succeeded
+      - failed
+      - canceled
+      - running
+    description: |
+      Describes the current state of an operation.
+```
+
+## Endpoint Descriptions
+
+### Start Operation
+
+Start an arbirary length operation.
+The response of the operation may be delivered synchronously (inline), or asynchronously, via a provided callback or the
+[Get Operation Result](#get-operation-result) endpoint.
+
+#### Paths
+
+- `/api/v1/services/{service}/operations/{operation}`
```

Comment:
This is what I had originally.
I buy removing the prefix, not sure about getting rid of (both) services and operations, it hurts extensibility.

Reply:
> it hurts extensibility.

Can by operation ID have slashes? If so, then any extensibility need we can just say "that's a built in operation".

Reply:
We might even want to take it to an extreme and make "service" a Temporal specific thing, the service is essentially the entire URL up to the operation.

Reply:
I think `/{service}/{operation}[/{sub-operation...]` makes sense

Reply:
@bergundy - If you want to upend it and make this fairly HTTP/URI oriented, yes we can have operation just be a potentially-containing-slashes name, so long as you can disambiguate the ID at the end which you should be able to.

So yes, if you want to take it to its extreme, everything is `/{operation}` and `operation` can be `/machines/gumballMachine/twist`. Unsure how far you want to go. I just don't think we need versioning or too much separation of concepts. But some is ok.

I don't have a strong opinion here. Reminder to others: this is a Nexus platform protocol API, we still expect users may expose HTTP for HTTP callers that then uses the protocol API, but you can design how your HTTP API there looks more.

Reply:
Hmm.. you can't have slashes in the operation name if the operation ID is optional.

Reply:
But I think changing this whole spec to say a Nexus endpoint may contain a URL prefix and there's no concept of a service in non-Temporal Nexus makes total sense.

Then all the endpoints become:
```
{http-prefix}/{operation}/{operation_id}
```

Reply:
Hrmm, there is likely value in a service abstraction same as there is for gRPC. You will have to come up with some name for that kind of layer when making the outbound Nexus call from a workflow. Because there is a server mapping. And remember, that outbound Nexus call needs to support non-Temporal one day.

Reply:
True,

We can map a service to a URL for Temporal.